File: web/src/api/sftp.ts
import { invoke } from '@tauri-apps/api/core'
import { invokeCommand, normalizeCommandError } from './tauri'
import { isMobileRuntime } from '@/lib/platform'
import type {
  SftpEntry,
  TransferTask,
  ConflictResolution,
  SftpCreateSessionResponse,
  SftpListResponse,
  SftpTreeResponse,
  SftpTransferResponse,
  SftpUploadResponse,
  SftpUploadBeginResponse,
  SftpDownloadResponse,
  SftpDeleteResponse,
  SftpFileReadResponse,
  SftpFileWriteRequest,
  SftpFileWriteResponse,
  DirectoryTransferMode,
  SftpMoveResponse,
} from '@/types/sftp'

const IPC_CHUNK_BYTES = 512 * 1024
const USE_BASE64_IPC = isMobileRuntime()

/** Rust SFTP session info. */
export interface SftpSessionInfo {
  id: string
  profile_id: string
  status: string
  error?: string
  home_dir?: string // User's home directory
  created_at: string
  host_key_fingerprint?: string
  known_host_key_fingerprint?: string
}

export const sftpApi = {
  // --- Session management ---

  createSession: (profileId: string) =>
    invokeCommand<SftpCreateSessionResponse>('sftp_create_session', { profileId }),

  getSession: (id: string) =>
    invokeCommand<SftpSessionInfo>('sftp_get_session', { id }),

  reconnectSession: (id: string) =>
    invokeCommand<SftpCreateSessionResponse>('sftp_reconnect_session', { id }),

  decideHostKey: (
    requestId: string,
    fingerprint: string,
    decision: 'trust_once' | 'trust_permanently' | 'reject',
  ) => invokeCommand<{ status: string; persisted?: boolean }>('sftp_host_key_decide', {
    requestId,
    fingerprint,
    decision,
  }),

  listSessions: () =>
    invokeCommand<SftpSessionInfo[]>('sftp_list_sessions'),

  closeSession: (id: string) =>
    invokeCommand<void>('sftp_close_session', { id }),

  // --- File operations ---

  list: (sessionId: string, path: string, showHidden = false) =>
    invokeCommand<SftpListResponse>('sftp_list', { sessionId, path, showHidden }),

  stat: (sessionId: string, path: string) =>
    invokeCommand<SftpEntry>('sftp_stat', { sessionId, path }),

  tree: (sessionId: string, path: string, depth = 3) =>
    invokeCommand<SftpTreeResponse>('sftp_tree', { sessionId, path, depth }),

  mkdir: (sessionId: string, path: string) =>
    invokeCommand<SftpEntry>('sftp_mkdir', { sessionId, path }),

  rename: (sessionId: string, oldPath: string, newPath: string) =>
    invokeCommand<SftpEntry>('sftp_rename', { sessionId, oldPath, newPath }),

  delete: (sessionId: string, paths: string[]) =>
    invokeCommand<SftpDeleteResponse>('sftp_delete', { sessionId, paths }),

  // --- Transfers ---

  /** Cross-session transfer: tries direct server-to-server copy first, falls
   *  back to backend relay if direct is not possible.
   *  If conflictResolution is "ask" (default) and the target has existing
   *  files, resolves with a `conflicts` array instead of starting a transfer.
   *  Pass "overwrite" | "rename" | "skip" to proceed without prompting. */
  transfer: async (
    sourceSessionId: string,
    targetSessionId: string,
    paths: string[],
    destDir: string,
    conflictResolution: ConflictResolution = 'ask',
    directoryMode: DirectoryTransferMode = 'archive',
  ): Promise<SftpTransferResponse> => {
    return invokeCommand<SftpTransferResponse>('sftp_transfer', {
        sourceSessionId,
        targetSessionId,
        paths,
        destDir,
        conflictResolution,
        directoryMode,
    })
  },

  move: async (
    sessionId: string,
    paths: string[],
    destDir: string,
    conflictResolution: ConflictResolution = 'ask',
  ): Promise<SftpMoveResponse> => {
    return invokeCommand<SftpMoveResponse>('sftp_move', {
      sessionId,
      paths,
      destDir,
      conflictResolution,
    })
  },

  /** Stream files through bounded raw IPC chunks. Each invoke completes only
   *  after Rust wrote the chunk, providing end-to-end backpressure. */
  upload: async (
    sessionId: string,
    files: File[],
    destDir: string,
    overwrite = false,
    onTasks?: (tasks: TransferTask[]) => void,
  ) => {
    const responses: SftpUploadResponse[] = []
    for (const file of files) {
      let uploadId: string | undefined
      try {
        const begin = await invokeCommand<SftpUploadBeginResponse>('sftp_upload_begin', {
          sessionId,
          name: file.name,
          destDir,
          overwrite,
          size: file.size,
        })
        uploadId = begin.upload_id
        onTasks?.(begin.tasks)
        const reader = file.stream().getReader()
        try {
          while (true) {
            const { done, value } = await reader.read()
            if (done) break
            for (let offset = 0; offset < value.byteLength; offset += IPC_CHUNK_BYTES) {
              const chunk = value.subarray(offset, offset + IPC_CHUNK_BYTES)
              if (USE_BASE64_IPC) {
                await invokeCommand('sftp_upload_chunk_base64', {
                  uploadId,
                  data: bytesToBase64(chunk),
                })
              } else {
                await invoke(
                  'sftp_upload_chunk',
                  chunk,
                  { headers: { 'x-eizhu-upload-id': uploadId } },
                )
              }
            }
          }
        } finally {
          reader.releaseLock()
        }
        responses.push(await invokeCommand<SftpUploadResponse>('sftp_upload_finish', { uploadId }))
      } catch (cause) {
        if (uploadId) {
          try {
            await invokeCommand<void>('sftp_upload_abort', { uploadId })
          } catch {
            // Preserve the original upload error.
          }
        }
        throw normalizeCommandError(cause)
      }
    }
    return { tasks: responses.flatMap((response) => response.tasks) }
  },

  uploadDocument: (
    sessionId: string,
    reference: string,
    destDir: string,
    overwrite = false,
  ) => invokeCommand<SftpUploadResponse>('sftp_upload_document', {
    sessionId,
    reference,
    destDir,
    overwrite,
  }),

  /** Stage remote files in Rust without loading them into memory. */
  download: (sessionId: string, paths: string[]) =>
    invokeCommand<SftpDownloadResponse>('sftp_download', { sessionId, paths }),

  exportDownload: (taskId: string) =>
    invokeCommand<string | null>('sftp_export_download', { taskId }),

  downloadToDocuments: async (sessionId: string, paths: string[]) => {
    const response = await invokeCommand<SftpDownloadResponse>('sftp_download', { sessionId, paths })
    for (const task of response.tasks) {
      while (true) {
        const current = (await invokeCommand<TransferTask[]>('sftp_list_transfers', { sessionId }))
          .find((candidate) => candidate.id === task.id)
        if (!current) throw new Error('下载任务已丢失')
        if (current.status === 'failed' || current.status === 'cancelled') {
          throw new Error(current.error_message || '下载未完成')
        }
        if (current.status === 'completed') break
        await new Promise((resolve) => setTimeout(resolve, 350))
      }
      await invokeCommand<string | null>('sftp_export_download', { taskId: task.id })
    }
    return response
  },

  /** Pull a completed staged download as a backpressured byte stream. */
  streamDownloadFile: (taskId: string): ReadableStream<Uint8Array> => {
    let offset = 0
    let closed = false
    const close = async () => {
      if (closed) return
      closed = true
      await invokeCommand<void>('sftp_download_close', { taskId })
    }
    return new ReadableStream<Uint8Array>({
      async pull(controller) {
        try {
          const chunk = USE_BASE64_IPC
            ? base64ToBytes(await invokeCommand<string>('sftp_download_chunk_base64', {
                taskId,
                offset,
                maxBytes: IPC_CHUNK_BYTES,
              }))
            : toUint8Array(await invoke<ArrayBuffer | Uint8Array>('sftp_download_chunk', {
                taskId,
                offset,
                maxBytes: IPC_CHUNK_BYTES,
              }))
          if (chunk.byteLength === 0) {
            controller.close()
            await close()
            return
          }
          offset += chunk.byteLength
          controller.enqueue(chunk)
        } catch (cause) {
          controller.error(normalizeCommandError(cause))
          try {
            await close()
          } catch {
            // Preserve the original download error.
          }
        }
      },
      async cancel() {
        await close()
      },
    })
  },

  listTransfers: (sessionId?: string, status?: string) => {
    return invokeCommand<TransferTask[]>('sftp_list_transfers', { sessionId, status })
  },

  cancelTransfer: (taskId: string) =>
    invokeCommand<{ id: string; status: string }>('sftp_cancel_transfer', { taskId }),

  clearCompletedTransfers: () =>
    invokeCommand<void>('sftp_clear_completed_transfers'),

  // --- Built-in editor ---

  /** Read a remote file as text for editing. Backend guards: >10MB → 413,
   *  binary → 415, non-UTF-8 → 415. Returns content + optimistic-lock token
   *  (mod_time) + Monaco language hint. */
  readFile: (sessionId: string, path: string) =>
    invokeCommand<SftpFileReadResponse>('sftp_read_file', { sessionId, path }),

  /** Write edited content back. Uses optimistic locking via expected_mod_time;
   *  mismatch → 409 FILE_MODIFIED. Returns the new mod_time for the next save. */
  writeFile: (sessionId: string, path: string, body: SftpFileWriteRequest) =>
    invokeCommand<SftpFileWriteResponse>('sftp_write_file', { sessionId, path, request: body }),
}

function toUint8Array(value: ArrayBuffer | Uint8Array): Uint8Array {
  return value instanceof Uint8Array ? value : new Uint8Array(value)
}

function bytesToBase64(bytes: Uint8Array): string {
  let binary = ''
  for (let offset = 0; offset < bytes.byteLength; offset += 32 * 1024) {
    binary += String.fromCharCode(...bytes.subarray(offset, offset + 32 * 1024))
  }
  return btoa(binary)
}

function base64ToBytes(value: string): Uint8Array {
  const binary = atob(value)
  const bytes = new Uint8Array(binary.length)
  for (let index = 0; index < binary.length; index += 1) {
    bytes[index] = binary.charCodeAt(index)
  }
  return bytes
}

File: src-tauri/src/commands/sftp.rs
//! Tauri IPC adapters for SFTP sessions, file operations and transfers.

use tauri::{ipc::InvokeBody, State};

use crate::{error::CommandError, sftp};

#[tauri::command]
pub(crate) async fn sftp_create_session(
    service: State<'_, sftp::SftpService>,
    profile_id: String,
) -> Result<sftp::SftpCreateSessionResponse, CommandError> {
    sftp::create_session(service.inner(), profile_id).await
}

#[tauri::command]
pub(crate) async fn sftp_get_session(
    service: State<'_, sftp::SftpService>,
    id: String,
) -> Result<sftp::SftpSessionInfo, CommandError> {
    sftp::get_session(service.inner(), id).await
}

#[tauri::command]
pub(crate) async fn sftp_reconnect_session(
    service: State<'_, sftp::SftpService>,
    id: String,
) -> Result<sftp::SftpCreateSessionResponse, CommandError> {
    service.reconnect(&id).await
}

#[tauri::command]
pub(crate) async fn sftp_host_key_decide(
    service: State<'_, sftp::SftpService>,
    request_id: String,
    fingerprint: String,
    decision: String,
) -> Result<serde_json::Value, CommandError> {
    service
        .decide_host_key(&request_id, fingerprint, &decision)
        .await
}

#[tauri::command]
pub(crate) async fn sftp_list_sessions(
    service: State<'_, sftp::SftpService>,
) -> Result<Vec<sftp::SftpSessionInfo>, CommandError> {
    sftp::list_sessions(service.inner()).await
}

#[tauri::command]
pub(crate) async fn sftp_close_session(
    service: State<'_, sftp::SftpService>,
    id: String,
) -> Result<(), CommandError> {
    sftp::close_session(service.inner(), id).await
}

#[tauri::command]
pub(crate) async fn sftp_list(
    service: State<'_, sftp::SftpService>,
    session_id: String,
    path: String,
    show_hidden: Option<bool>,
) -> Result<sftp::SftpListResponse, CommandError> {
    sftp::list(service.inner(), session_id, path, show_hidden).await
}

#[tauri::command]
pub(crate) async fn sftp_stat(
    service: State<'_, sftp::SftpService>,
    session_id: String,
    path: String,
) -> Result<sftp::SftpEntry, CommandError> {
    sftp::stat(service.inner(), session_id, path).await
}

#[tauri::command]
pub(crate) async fn sftp_tree(
    service: State<'_, sftp::SftpService>,
    session_id: String,
    path: String,
    depth: Option<u32>,
) -> Result<sftp::SftpTreeResponse, CommandError> {
    sftp::tree_entries(service.inner(), session_id, path, depth).await
}

#[tauri::command]
pub(crate) async fn sftp_mkdir(
    service: State<'_, sftp::SftpService>,
    session_id: String,
    path: String,
) -> Result<sftp::SftpEntry, CommandError> {
    sftp::mkdir(service.inner(), session_id, path).await
}

#[tauri::command]
pub(crate) async fn sftp_rename(
    service: State<'_, sftp::SftpService>,
    session_id: String,
    old_path: String,
    new_path: String,
) -> Result<sftp::SftpEntry, CommandError> {
    sftp::rename(service.inner(), session_id, old_path, new_path).await
}

#[tauri::command]
pub(crate) async fn sftp_delete(
    service: State<'_, sftp::SftpService>,
    session_id: String,
    paths: Vec<String>,
) -> Result<sftp::SftpDeleteResponse, CommandError> {
    sftp::delete(service.inner(), session_id, paths).await
}

#[tauri::command]
pub(crate) async fn sftp_read_file(
    service: State<'_, sftp::SftpService>,
    session_id: String,
    path: String,
) -> Result<sftp::SftpFileReadResponse, CommandError> {
    sftp::read_file(service.inner(), session_id, path).await
}

#[tauri::command]
pub(crate) async fn sftp_write_file(
    service: State<'_, sftp::SftpService>,
    session_id: String,
    path: String,
    request: sftp::SftpFileWriteRequest,
) -> Result<sftp::SftpFileWriteResponse, CommandError> {
    sftp::write_file(service.inner(), session_id, path, request).await
}

#[tauri::command]
pub(crate) async fn sftp_upload_begin(
    service: State<'_, sftp::SftpService>,
    session_id: String,
    name: String,
    dest_dir: String,
    overwrite: bool,
    size: u64,
) -> Result<sftp::SftpUploadBeginResponse, CommandError> {
    sftp::sftp_upload_begin(service.inner(), session_id, name, dest_dir, overwrite, size).await
}

#[tauri::command]
pub(crate) async fn sftp_upload_chunk(
    service: State<'_, sftp::SftpService>,
    request: tauri::ipc::Request<'_>,
) -> Result<sftp::SftpUploadChunkResponse, CommandError> {
    let upload_id = request
        .headers()
        .get("x-eizhu-upload-id")
        .and_then(|value| value.to_str().ok())
        .ok_or_else(|| CommandError::new("VALIDATION", "upload id is required"))?;
    let InvokeBody::Raw(bytes) = request.body() else {
        return Err(CommandError::new(
            "INVALID_FORM",
            "raw upload chunk is required",
        ));
    };
    sftp::upload_chunk(service.inner(), upload_id, bytes).await
}

#[tauri::command]
pub(crate) async fn sftp_upload_chunk_base64(
    service: State<'_, sftp::SftpService>,
    upload_id: String,
    data: String,
) -> Result<sftp::SftpUploadChunkResponse, CommandError> {
    sftp::sftp_upload_chunk_base64(service.inner(), upload_id, data).await
}

#[tauri::command]
pub(crate) async fn sftp_upload_finish(
    service: State<'_, sftp::SftpService>,
    upload_id: String,
) -> Result<sftp::SftpUploadResponse, CommandError> {
    sftp::sftp_upload_finish(service.inner(), upload_id).await
}

#[tauri::command]
pub(crate) async fn sftp_upload_abort(
    service: State<'_, sftp::SftpService>,
    upload_id: String,
) -> Result<(), CommandError> {
    sftp::sftp_upload_abort(service.inner(), upload_id).await
}

#[tauri::command]
pub(crate) async fn sftp_download(
    service: State<'_, sftp::SftpService>,
    session_id: String,
    paths: Vec<String>,
) -> Result<sftp::SftpDownloadResponse, CommandError> {
    sftp::sftp_download(service.inner(), session_id, paths).await
}

#[tauri::command]
pub(crate) async fn sftp_download_chunk(
    service: State<'_, sftp::SftpService>,
    task_id: String,
    offset: u64,
    max_bytes: u32,
) -> Result<tauri::ipc::Response, CommandError> {
    let bytes = sftp::sftp_download_chunk(service.inner(), task_id, offset, max_bytes).await?;
    Ok(tauri::ipc::Response::new(bytes))
}

#[tauri::command]
pub(crate) async fn sftp_download_chunk_base64(
    service: State<'_, sftp::SftpService>,
    task_id: String,
    offset: u64,
    max_bytes: u32,
) -> Result<String, CommandError> {
    sftp::sftp_download_chunk_base64(service.inner(), task_id, offset, max_bytes).await
}

#[tauri::command]
pub(crate) async fn sftp_download_close(
    service: State<'_, sftp::SftpService>,
    task_id: String,
) -> Result<(), CommandError> {
    sftp::sftp_download_close(service.inner(), task_id).await
}

#[tauri::command]
pub(crate) async fn sftp_list_transfers(
    service: State<'_, sftp::SftpService>,
    session_id: Option<String>,
    status: Option<String>,
) -> Result<Vec<sftp::TransferTask>, CommandError> {
    sftp::sftp_list_transfers(service.inner(), session_id, status).await
}

#[tauri::command]
pub(crate) async fn sftp_cancel_transfer(
    service: State<'_, sftp::SftpService>,
    task_id: String,
) -> Result<serde_json::Value, CommandError> {
    sftp::sftp_cancel_transfer(service.inner(), task_id).await
}

#[tauri::command]
pub(crate) async fn sftp_clear_completed_transfers(
    service: State<'_, sftp::SftpService>,
) -> Result<(), CommandError> {
    sftp::sftp_clear_completed_transfers(service.inner()).await
}

#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub(crate) async fn sftp_transfer(
    service: State<'_, sftp::SftpService>,
    source_session_id: String,
    target_session_id: String,
    paths: Vec<String>,
    dest_dir: String,
    conflict_resolution: Option<String>,
    directory_mode: Option<String>,
) -> Result<sftp::SftpTransferResponse, CommandError> {
    sftp::sftp_transfer(
        service.inner(),
        source_session_id,
        target_session_id,
        paths,
        dest_dir,
        conflict_resolution,
        directory_mode,
    )
    .await
}

#[tauri::command]
pub(crate) async fn sftp_move(
    service: State<'_, sftp::SftpService>,
    session_id: String,
    paths: Vec<String>,
    dest_dir: String,
    conflict_resolution: Option<String>,
) -> Result<sftp::SftpMoveResponse, CommandError> {
    sftp::sftp_move(
        service.inner(),
        session_id,
        paths,
        dest_dir,
        conflict_resolution,
    )
    .await
}
